src/session.c

```diff
@@ -395,24 +395,22 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
 
     token = p11prov_slot_get_token(slot);
     if (!(token->flags & CKF_PROTECTED_AUTHENTICATION_PATH)) {
+        const char *cached_pin = p11prov_slot_get_cached_pin(slot);
         const char *bad_pin = p11prov_slot_get_bad_pin(slot);
         if (uri) {
             pin = (CK_UTF8CHAR_PTR)p11prov_uri_get_pin(uri);
-            if (bad_pin && strcmp((const char *)pin, bad_pin) == 0) {
-                P11PROV_raise(session->provctx, CKR_PIN_INVALID,
-                              "Blocking stored PIN that failed a previous login"
-                              " to avoid blocking the token");
-                pin = NULL;
-            }
         }
         if (!pin) {
             pin = p11prov_ctx_pin(session->provctx);
-            if (bad_pin && strcmp((const char *)pin, bad_pin) == 0) {
-                P11PROV_raise(session->provctx, CKR_PIN_INVALID,
-                              "Blocking stored PIN that failed a previous login"
-                              " to avoid blocking the token");
-                pin = NULL;
-            }
+        }
+        if (!pin && cached_pin) {
+            pin = (CK_UTF8CHAR_PTR)cached_pin;
+        }
+        if (pin && bad_pin && strcmp((char *)pin, bad_pin) == 0) {
+            P11PROV_raise(session->provctx, CKR_PIN_INVALID,
+                          "Blocking stored PIN that failed a previous login"
+                          " to avoid blocking the token");
```

Comment:
maybe update this log message that it can now affects also cached pin.

Reply:
I am not sure why, if the pin fails it is never cached in the first place.